### PR TITLE
fix(core/issue-notes): Make issue-summary section more modular

### LINKED ISSUE #2207: Handling of <section id="issue-summary"> is poor
## Important info

* URL to affected spec: (not online yet) 
* ReSpec version: 24.7.0
* [x] I did a "hard refresh", but it's still busted. 

## Description of problem

ReSpec doesn't work well if you try to customize the Issue Summary.

According to [the wiki on `issue-summary`](https://github.com/w3c/respec/wiki/issue-summary),
if you create a section with that `id`, it will be used for the auto-generated issue summary.  Like similar auto-generated sections, I assumed that meant I could add my own heading and intro paragraph.

**What happened?:** 

The auto generated issue summary generates and adds it's own heading even if I've included one, so I end up with two `<h2>` elements in the section.

If I _don't_ include my own heading in the section, I get the "no-headingless-sections" linter error!

**Minimal test case (for the body of a generic ReSpec document):**

```html
<p class="issue">Write the rest of this spec!</p>

<section class="appendix" id="issue-summary">
<h2>Open Issues</h2><!-- or remove this to see the linter error -->
<p>
We are tracking issues on
<a href="https://github.com/w3c/respec/issues">Github</a>.
</p>
</section>
```

### PR DESCRIPTION
**Pull request by pradeepgangwar**
fixes #2207

Works as follows:
- If heading and paragraph is provided it uses both of them and appends issues list at the end.
- If only paragraph is given it uses default title and appends issues list after adding paragraph.
- If nothing is give then it chooses default title and appends issues list.

### CODE REVIEW

**Comment by pradeepgangwar** [comment]
@marcoscaceres Working as expected. Added few tests to check functionality. Please review.

File: src/core/issues-notes.js [hunk 1]
```diff
@@ -133,7 +133,29 @@ function handleIssues(ins, ghIssues, conf) {
     }
     pub(report.type, report);
   });
-  createIssueSummarySection(l10n.issue_summary, issueList);
+  const issueSummaryElement = document.getElementById("issue-summary");
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: Can you help me understand this change a bit? I'm not sure why inline this is different/better than the original code?

Comment (marcoscaceres) [reply src/core/issues-notes.js]: Oh, sorry, I got confused... I was looking at a subset of changes. However, having this in a function makes sense. This code block is getting rather long.

**Comment by pradeepgangwar** [comment]
Did the changes. Please review. :)

File: src/core/issues-notes.js [hunk 2]
```diff
@@ -203,6 +193,36 @@ function createIssueSummaryEntry(l10nIssue, report, id) {
...
 }
 
+/**
+ *
+ * @param {*} issueList
+ */
+function makeIssueSectionSummary(issueList) {
+  const issueSummaryElement = document.getElementById("issue-summary");
+
+  if (issueSummaryElement) {
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: Avoid a level of nesting: 
```suggestion
  if (!issueSummaryElement) return;
```

File: src/core/issues-notes.js [hunk 3]
```diff
@@ -203,6 +193,36 @@ function createIssueSummaryEntry(l10nIssue, report, id) {
...
+        issueSummaryElement.removeChild(paragraph);
+        const issueSummary = hyperHTML`
+          <div><h2>${l10n.issue_summary}</h2>${paragraph}${issueList}</div>`;
+        issueSummaryElement.append(...issueSummary.childNodes);
+      } else {
+        const issueSummary = hyperHTML`
+          <div><h2>${l10n.issue_summary}</h2>${issueList}</div>`;
+        issueSummaryElement.append(...issueSummary.childNodes);
+      }
+    } else {
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: I'm thinking that this is no longer an error. If the `issueList` doesn't have any children, then it should say: `<p>There are no issues listed in this specification.</p>`. The string should be localizable, but we can get to that.

File: src/core/issues-notes.js [hunk 4]
```diff
@@ -203,6 +193,36 @@ function createIssueSummaryEntry(l10nIssue, report, id) {
...
 
+/**
+ *
+ * @param {*} issueList
+ */
+function makeIssueSectionSummary(issueList) {
+  const issueSummaryElement = document.getElementById("issue-summary");
+
+  if (issueSummaryElement) {
+    if (document.querySelectorAll(".issue").length) {
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: Should really use `issueList` here... not query the document.

File: src/core/issues-notes.js [hunk 5]
```diff
@@ -203,6 +193,36 @@ function createIssueSummaryEntry(l10nIssue, report, id) {
...
+      const heading = issueSummaryElement.querySelector("h2");
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: Note that  `issue-summary` can appear anywhere, and hence there is no guarantee that the first heading will be a h2. It's perfectly valid for it to be any heading level 2-6.

File: src/core/issues-notes.js [hunk 6]
```diff
@@ -203,6 +193,36 @@ function createIssueSummaryEntry(l10nIssue, report, id) {
...
+      const paragraph = issueSummaryElement.querySelector("p");
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: Likewise, there is no restriction on what content there can be inside the summary:

```HTML
<section>
   <h3>Appendix</h3>
   <section id="issue-summary">
     <h4>OMG so many issues!</h4>
     <aside class="issue">Do we have too many issues?</aside>
     <figure><figcaption>A diagram</figcaption></figure>
   </section>
</section> 
```

File: src/core/issues-notes.js [hunk 7]
```diff
@@ -203,6 +193,36 @@ function createIssueSummaryEntry(l10nIssue, report, id) {
...
+      if (heading && paragraph) {
+        issueSummaryElement.append(issueList);
+      } else if (!heading && paragraph) {
+        issueSummaryElement.removeChild(paragraph);
+        const issueSummary = hyperHTML`
+          <div><h2>${l10n.issue_summary}</h2>${paragraph}${issueList}</div>`;
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: We should ditch the `<div>`

File: src/core/issues-notes.js (hunk 3 above)
Comment (pradeepgangwar) [reply src/core/issues-notes.js]: I figured out that function `handleIssues` is not called when there are no issues. Hence the function that i created is never called thus never providing me opportunity to show the error `<p>There are no issues listed in this specification.</p>`.

Check:

Comment (pradeepgangwar) [reply src/core/issues-notes.js]: If you want a different behaviour we will probably have to change this. I am not sure how to change this to meet ReSpec requirements.

File: src/core/issues-notes.js [hunk 8]
```diff
@@ -203,6 +193,30 @@ function createIssueSummaryEntry(l10nIssue, report, id) {
...
+  const issueSummaryElement = document.getElementById("issue-summary");
+
+  if (!issueSummaryElement) return;
+  const heading = issueSummaryElement.querySelectorAll("h2, h3, h4, h5, h6");
+  const defaultHeading = hyperHTML`<h2>${l10n.issue_summary}</h2>`;
+
+  issueList.hasChildNodes()
+    ? issueSummaryElement.append(issueList)
+    : issueSummaryElement.append(
+        hyperHTML`<p>There are no issues listed in this specification.</p>`
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: We should localize this string. Please add it to `localizationStrings` above as "no_issues_in_spec". 

```suggestion
        hyperHTML`<p>${l10n.no_issues_in_spec}</p>`
```

For Spanish, please use: 

"No hay problemas enumerados en esta especificación." 

@thijsbrentjens, could you suggest a Dutch translation for: 

"There are no issues listed in this specification."

File: tests/spec/core/issues-notes-spec.js [hunk 9]
```diff
@@ -378,4 +378,47 @@ describe("Core — Issues and Notes", () => {
...
+  it("shows issue-summary section with paragraph and default heading when only <p> is defined", async () => {
+    const ops = {
+      config: makeBasicConfig(),
+      body: `
+      <section>
+        <h2>Test Issues</h2>
+        <p class="issue" data-number=123></p>
+      </section>
+      <section id="issue-summary">
+        <p>Here you will find all issues summary</p>
```
Comment (marcoscaceres) [inline tests/spec/core/issues-notes-spec.js]: I think we should add one more random element, like a `<div class="note">`, just to make sure it's also being added.

File: src/core/issues-notes.js [hunk 10]
```diff
@@ -203,6 +193,30 @@ function createIssueSummaryEntry(l10nIssue, report, id) {
...
 
+/**
+ *
+ * @param {*} issueList
+ */
+function makeIssueSectionSummary(issueList) {
+  const issueSummaryElement = document.getElementById("issue-summary");
+
+  if (!issueSummaryElement) return;
+  const heading = issueSummaryElement.querySelectorAll("h2, h3, h4, h5, h6");
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: I think here you just want whatever the first one is. The assumption being that there is a heading in the section, which is fine. 
 
```suggestion
  const heading = issueSummaryElement.querySelector("h2, h3, h4, h5, h6");
```

File: src/core/issues-notes.js [hunk 11]
```diff
@@ -203,6 +193,30 @@ function createIssueSummaryEntry(l10nIssue, report, id) {
...
+  const defaultHeading = hyperHTML`<h2>${l10n.issue_summary}</h2>`;
+
+  issueList.hasChildNodes()
+    ? issueSummaryElement.append(issueList)
+    : issueSummaryElement.append(
+        hyperHTML`<p>There are no issues listed in this specification.</p>`
+      );
+  heading.length
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]:
```suggestion
  heading
```

File: src/core/issues-notes.js [hunk 12]
```diff
@@ -203,6 +193,30 @@ function createIssueSummaryEntry(l10nIssue, report, id) {
...
+ */
+function makeIssueSectionSummary(issueList) {
+  const issueSummaryElement = document.getElementById("issue-summary");
+
+  if (!issueSummaryElement) return;
+  const heading = issueSummaryElement.querySelectorAll("h2, h3, h4, h5, h6");
+  const defaultHeading = hyperHTML`<h2>${l10n.issue_summary}</h2>`;
+
+  issueList.hasChildNodes()
+    ? issueSummaryElement.append(issueList)
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: I'm a bit worried that this `issueSummaryElement.append(issueList)` appears twice (here and above)? I was only expecting to see that once.

File: src/core/issues-notes.js (hunk 8 above)
Comment (pradeepgangwar) [reply src/core/issues-notes.js]: I did a google translate for this. Let me know if that needs some changes.

File: src/core/issues-notes.js (hunk 12 above)
Comment (pradeepgangwar) [reply src/core/issues-notes.js]: I noticed that and tested it for various cases and it works fine. It does not duplicate the content. Should I do it another way?

**Comment by pradeepgangwar** [comment]
I have made the changes. I have also added some inline comments. Let me know if we need more changes. :)

File: src/core/issues-notes.js [hunk 13]
```diff
@@ -22,12 +22,15 @@ export const name = "core/issues-notes";
 const localizationStrings = {
   en: {
     issue_summary: "Issue Summary",
+    no_issues_in_spec: "There are no issues listed in this specification.",
   },
   nl: {
     issue_summary: "Lijst met issues",
+    no_issues_in_spec: "No hay problemas enumerados en esta especificación.",
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: This is spanish ;)

File: src/core/issues-notes.js [hunk 14]
```diff
@@ -22,12 +22,15 @@ export const name = "core/issues-notes";
...
   },
   es: {
     issue_summary: "Resumen de la cuestión",
+    no_issues_in_spec: "Er zijn geen problemen vermeld in deze specificatie.",
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: Please be sure @thijsbrentjens verifies this translation. Using auto-translations is always a big risk.

File: src/core/issues-notes.js (hunk 12 above)
Comment (marcoscaceres) [reply src/core/issues-notes.js]: yeah, please see if there is a way that `issueSummaryElement.append(issueList)` is only ever called once.

File: src/core/issues-notes.js [hunk 15]
```diff
@@ -203,6 +196,26 @@ function createIssueSummaryEntry(l10nIssue, report, id) {
...
+  const issueSummaryElement = document.getElementById("issue-summary");
+
+  if (!issueSummaryElement) return;
+  const heading = issueSummaryElement.querySelector("h2, h3, h4, h5, h6");
+
+  issueList.hasChildNodes()
+    ? issueSummaryElement.append(issueList)
+    : issueSummaryElement.append(hyperHTML`<p>${l10n.no_issues_in_spec}</p>`);
+  heading ||
+    issueSummaryElement.insertBefore(
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]: Just using `if (!heading) {` would be cleaner. 

I still think there is a bug tho... you could have:

```HTML
<section id="issue-summary">
  <p>Issues are amazing</p>
  <section>
     <h3>Amazing</h3>
  </section>
</section>
```

(Make sure you add the above to a test)

So there is check needed: that the `heading !== issueSummaryElement.firstElementChild`.

Comment (marcoscaceres) [reply src/core/issues-notes.js]: I wonder if `insertAdjacentHTML("afterbegin”, ...)` couldn’t be used here instead 🧐 if yes, then it’s good as it avoids having to use hyperHTML).

**Comment by pradeepgangwar** [comment]
@marcoscaceres Replaced `insertAfter` with `inserAdjecantHTML` and added tests for below case: 😃 

```
<section id="issue-summary">
  <p>Issues are amazing</p>
  <section>
     <h3>Amazing</h3>
  </section>
</section>
```

File: src/core/issues-notes.js [hunk 16]
```diff
@@ -203,6 +196,30 @@ function createIssueSummaryEntry(l10nIssue, report, id) {
   `;
 }
 
+/**
+ *
+ * @param {*} issueList
```
Comment (marcoscaceres) [inline src/core/issues-notes.js]:
```suggestion
 * @param {HTMLUListElement} issueList
```

**Comment by pradeepgangwar** [comment]
@marcoscaceres Undoing commit https://github.com/w3c/respec/pull/2219/commits/5840790024e57a108dfaebf15132d39a3284fcae makes all the tests pass. 😕 
I don't know why that happens.

**Comment by marcoscaceres** [comment]
Oops, I broke it. Will fix.

**Comment by marcoscaceres** [comment]
Sorry about that @pradeepgangwar ... was distracted in meeting and "fixing" things in the GitHub editor... that always ends badly :)

**Pull request merged**